src/plugins/config_reader.go

```diff
@@ -67,33 +68,71 @@ func (r *ConfigReader) Process(msg *core.Message) {
 	switch cmd := msg.Data().(type) {
 	case *proto.Command:
 		switch msg.Topic() {
-		case core.AgentConfig:
+		case core.AgentConfig, core.AgentConnected:
 			// Update the agent config on disk
-			r.updateAgentConfig(cmd)
+			switch commandData := cmd.Data.(type) {
+			case *proto.Command_AgentConfig:
+				r.updateAgentConfig(commandData.AgentConfig)
+			case *proto.Command_AgentConnectResponse:
+				r.updateAgentConfig(commandData.AgentConnectResponse.AgentConfig)
+			}
 		}
 	}
 }
 
 func (r *ConfigReader) Subscriptions() []string {
-	return []string{core.CommMetrics, core.AgentConfig, core.AgentConfigChanged}
+	return []string{core.CommMetrics, core.AgentConfig, core.AgentConfigChanged, core.AgentConnected}
 }
 
-func (r *ConfigReader) updateAgentConfig(cmd *proto.Command) {
-	switch commandData := cmd.Data.(type) {
-	case *proto.Command_AgentConfig:
-		configUpdated, err := config.UpdateAgentConfig(r.config.ClientID, commandData.AgentConfig.Details.Tags, commandData.AgentConfig.Details.Features)
-		if err != nil {
-			log.Errorf("Failed updating Agent config - %v", err)
+func (r *ConfigReader) updateAgentConfig(payloadAgentConfig *proto.AgentConfig) {
+	if payloadAgentConfig != nil && payloadAgentConfig.Details != nil {
+
+		onDiskAgentConfig, _ := config.GetConfig(r.config.ClientID)
+		synchronizeFeatures := false
+		synchronizeTags := false
+
+		if payloadAgentConfig.Details.Features != nil {
+
+			for index, feature := range payloadAgentConfig.Details.Features {
+				payloadAgentConfig.Details.Features[index] = strings.Replace(feature, "features_", "", 1)
+			}
+
+			sort.Strings(onDiskAgentConfig.Features)
+			sort.Strings(payloadAgentConfig.Details.Features)
+			synchronizeFeatures = !reflect.DeepEqual(payloadAgentConfig.Details.Features, onDiskAgentConfig.Features)
+
+		} else {
+			payloadAgentConfig.Details.Features = onDiskAgentConfig.Features
+		}
+
+		if payloadAgentConfig.Details.Tags != nil {
+			sort.Strings(onDiskAgentConfig.Tags)
+			sort.Strings(payloadAgentConfig.Details.Tags)
+			synchronizeTags = !reflect.DeepEqual(payloadAgentConfig.Details.Tags, onDiskAgentConfig.Tags)
 		}
 
-		// If the config was updated send a new agent config updated message
-		if configUpdated {
-			log.Debugf("Updated agent config on disk")
-			r.messagePipeline.Process(core.NewMessage(core.AgentConfigChanged, ""))
+		if payloadAgentConfig.Details.Server == nil || payloadAgentConfig.Details.Server.Backoff == nil {
```

Comment:
I think you can remove this if statement. If there are no Backoff settings in the payloadAgentConfig then we shouldn't be updating the settings

Reply:
Currently if the backoff settings are nil they get set to a default, was trying to keep the same behaviour but can change it if that behaviour was not correct https://github.com/nginx/agent/blob/e59c2a3fdda2175ef15c34c31c293d8df4fe0bbe/src/plugins/commander.go#L86

Reply:
Ok lets keep it as it is for now. I'll address the issue in a separate PR.